_posts/2021-02-26-dune-3-plan.md

```diff
@@ -0,0 +1,87 @@
+---
+layout: blog
+title: Plan for Dune 3.0
+author: andreymokhov
+tags: [dune]
+---
+
+<!-- TODO: After publishing the post, create a Discuss page, and link to it. -->
+
+The Dune team has started planning the next major release and we would like to
+share and discuss the current plan with the community. We expect to release
+Dune 3 in August 2021. As with previous major releases, we will put a lot of
+effort in backward compatibility, which means the majority of Dune users should
+not notice any changes.
+
+## What's new
+
+Here are the main features we plan include in Dune 3:
+
+* **Incremental file-watching mode.** Dune already supports a basic
+  file-watching mode, activated by the `-w` flag, where the build is restarted
+  whenever any source file changes. The current implementation is
+  non-incremental: restarting really means restarting the build from scratch,
+  which, for example, involves reparsing all `dune` files. On small-size
+  projects, this is fast enough, but this clearly doesn't scale well to large
+  monorepos. We are making the file-watching mode fully incremental, to avoid
+  redoing the computations whose inputs didn't change since the previous build.
+
+* **Dune RPC.** We are working on adding an RPC server to Dune that will make it
+  possible to control Dune programmatically (instead of using the command line),
+  which is often desirable for integration with various other tools, such as
+  editors.
+
+* **Better scoping semantics.** Dune provides some project-level scoping
+  functionality, for example, it's possible to mark a library as private to
+  prevent leaking the implementation details outside of a project. The current
+  implementation of scoping has wrong semantics for nested vendored projects: a
+  public library in the inner vendored project escapes its scope and is globally
+  visible. This loophole caused problems for users, and we are going to remove
+  it. This will be a breaking change but nested vendored projects are not very
+  common, so this change will only affect a small number of users and in a
+  positive way.
+
```

Comment:
The scoping semantics are the only breaking change from a dune-lang perspective and may be worth pulling out specially to explain the whole monorepo situation. The others could all (in theory) be put into an incremental release of dune 2.x.

Reply:
I'm a bit hesitant to put the emphasis on scoping semantics as the only breaking change. The change in warnings, for example, might break things too (if a script runs `dune` and expects empty output). Similar for some other changes.

The new intro says that some changes cannot be guarded by the Dune language version, and I think scoping semantics is one such change (but not the only one).

Reply:
> The change in warnings, for example, might break things too (if a script runs dune and expects empty output). Similar for some other changes.

This kind of breaking changes are less dramatic. Dune should be able to work on a workspace composed of different projects developed by different people and using different versions of the language. This is why backward compatibility at the language level is very important. For the rest, such as the cli behaviour, this is still important but we can allow ourselves a bit more freedom to change things at each major release. The only cli invocation we should be particularly mindful about is `dune build -p <package>`, because that's the one we tell people to use in their opam file.

Reply:
I added a separate section on breaking changes with respect to the language, adding one more change that we previously overlooked. Please have a look.